src/Network/Riak/Cluster.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE DeriveDataTypeable    #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+module Network.Riak.Cluster
+    ( Cluster(..)
+    , InClusterError(..)
+    , connectToCluster
+    , inCluster
+    , Riak.create
+    , Riak.defaultClient
+    ) where
+
+import           Control.Exception
+import           Control.Exception.Enclosed
+import           Control.Monad.Catch           (MonadThrow (..))
+import           Control.Monad.Trans.Control   (MonadBaseControl)
+import           Data.Typeable
+import           Data.Vector                   (Vector)
+import qualified Data.Vector                   as V
+import           Network.Riak                  (Connection)
+import qualified Network.Riak                  as Riak
+import qualified Network.Riak.Connection.Pool  as Riak
+import           System.Random.Mersenne.Pure64
+import           System.Random.Shuffle         (shuffle')
+
+-- | Datatype holding connection-pool with all known cluster nodes
+data Cluster = Cluster
+    { clusterPools :: Vector Riak.Pool
+      -- ^ Vector of connection pools to riak cluster nodes
+    , clusterGen   :: PureMT
+    }
+    deriving (Show)
+
+-- | Error that gets thrown whenever operation couldn't succeed with
+-- any node.
+data InClusterError = InClusterError [SomeException]
+    deriving (Show, Typeable)
+instance Exception InClusterError
+
+-- | Function to connect to riak cluster with sane pool defaults
+connectToCluster :: [Riak.Client] -> IO Cluster
+connectToCluster clients = do
+    pools <- mapM (\c -> Riak.create c 1 10 20) clients
+    connectToClusterWithPools pools
+
+-- | Function to connect to riak cluster with pre-created list of
+-- 'Riak.Pool' objects
+connectToClusterWithPools :: [Riak.Pool] -> IO Cluster
+connectToClusterWithPools pools = do
+    mt <- newPureMT
+    return (Cluster (V.fromList pools) mt)
+
+-- | Tries to run some operation for a random riak node. If it fails,
+-- tries all other nodes. If all other nodes fail - throws
+-- 'InClusterError' exception.
+inCluster :: (MonadThrow m, MonadBaseControl IO m)
+          => Cluster -> (Connection -> m a) -> m a
+inCluster rc f = do
+    let pools = shuffle' (V.toList (clusterPools rc))
+                         (V.length (clusterPools rc))
+                         (clusterGen rc)
+    go pools []
+  where
+    go [] errors = throwM (InClusterError errors)
+    go (p:ps) es = Riak.withConnectionM p $ \c -> do
+        er <- tryAny (f c)
+        either (\err -> go ps (err:es))
+               return er
```

Comment:
@tmcgilchrist I like this idea, but I would do that as a separate issue.

First, I think it might be a better idea to keep `inCluster` as is, and have separate `inClusterExt`, which would be parameterized by strategy. Current riak driver already has these `Default Default Default` APIs, and I'd be happy to think on possible solutions to have some `Network.Riak.Simple` API, suitable for a "Hello world" example, which could look as simple as ones for other languages (e.g., just do the obvious default, without having to even read about strategies and machine-confirmation-numbers).

Another problem to think of – maybe user would want to set retry-strategy inside a `Cluster` config itself upon connection, not at request time. I think this would be useful for those who want to control retry-strategy.

And last point to address, I'm not sure `Control.Retry` is the right thing to go here. That package seems to work for functions which retry doing the same action with some delay number, set by strategy. With riak driver, what we want to handle (in most cases) is a situation when one of cluster nodes is down. If it's down – you'd usually want to immediately speak to another node, e.g. have no observable affect to your program's operation abilities when one (or several) of nodes is down. Maybe exception-catching could be better done for specific types of errors, if I'll know which ones – I'll make another PR for that. If we'd want to provide some precise control on retry strategies – we'd need something, which describes not only timeouts, but also a way to set how much more nodes to try, and which ones exactly.

To repeat my point, I'd go with separate issue and possibly separate extended function if we'd want this. Please tell what you think on this!